components/category_menu_items/category_menu_items.tsx

```diff
@@ -0,0 +1,115 @@
+    }, []);
+
+    const moveToNewCategory = useCallback(() => {
+        dispatch(openModal({
+            modalId: ModalIdentifiers.EDIT_CATEGORY,
+            dialogType: EditCategoryModal,
+            dialogProps: {
+                channelIdsToAdd: multiSelectedChannelIds.indexOf(channel.id) === -1 ? [channel.id] : multiSelectedChannelIds,
+            },
+        }));
+        trackEvent('ui', 'ui_sidebar_channel_menu_createCategory');
+    }, []);
+
+    const selectedChannels = multiSelectedChannelIds.indexOf(channel.id) === -1 ? [channel] : displayedChannels.filter((c) => multiSelectedChannelIds.indexOf(c.id) !== -1);
+    const allChannelsAreDMs = selectedChannels.every((selectedChannel) => selectedChannel.type === Constants.DM_CHANNEL || selectedChannel.type === Constants.GM_CHANNEL);
+    const allChannelsAreNotDMs = selectedChannels.every((selectedChannel) => selectedChannel.type !== Constants.DM_CHANNEL && selectedChannel.type !== Constants.GM_CHANNEL);
+
+    const categoryMenuItems = categories?.filter((category) => {
+        if (allChannelsAreDMs) {
+            return category.type !== CategoryTypes.CHANNELS;
+        } else if (allChannelsAreNotDMs) {
+            return category.type !== CategoryTypes.DIRECT_MESSAGES;
+        }
+        return true;
+    }).map((category: ChannelCategory) => {
+        return {
+            id: `moveToCategory-${channel.id}-${category.id}`,
+            icon: category.type === CategoryTypes.FAVORITES ? (<i className='icon-star-outline'/>) : (<i className='icon-folder-outline'/>),
+            direction: 'right' as any,
+            text: category.display_name,
+            action: () => moveToCategory(category.id),
+        } as any;
+    });
+
+    categoryMenuItems?.push(
+        {
+            id: 'ChannelMenu-moveToDivider',
+            text: (<li className='MenuGroup menu-divider'/>),
+        },
+        {
+            id: `moveToNewCategory-${channel.id}`,
+            icon: (<i className='icon-folder-move-outline'/>),
+            direction: 'right' as any,
+            text: intl.formatMessage({id: 'sidebar_left.sidebar_channel_menu.moveToNewCategory', defaultMessage: 'New Category'}),
+            action: moveToNewCategory,
+        },
+    );
+
+    return categories ? (
+        <React.Fragment>
+            <Menu.Group>
+                <Menu.ItemSubMenu
+                    id={`moveTo-${channel.id}`}
+                    subMenu={categoryMenuItems}
+                    text={intl.formatMessage({id: 'sidebar_left.sidebar_channel_menu.moveTo', defaultMessage: 'Move to...'})}
+                    icon={<i className='icon-folder-move-outline'/>}
+                    direction={'right' as any}
+                    openUp={openUp}
+                    styleSelectableItem={true}
+                    selectedValueText={currentCategory?.display_name}
+                    renderSelected={false}
+                />
+            </Menu.Group>
+        </React.Fragment>
+    ) : null;
```

Comment:
Nit: 0/5, but I think for improved clarity, we could check if categories is falsy and return null before we return the actual component.